### ISSUE TITLE
What to do after training

### ISSUE THREAD

**Issue by CesMak** [henrycharlesworth/big2_PPOalgorithm#8]
Hey Henry,

I would like to work with your code. For this I would like to ask:

The Training is creating a modelParameters0, modelParameters500, etc. files. How can I now use this file and test if my training really worked? Could you give a minimal short example for this?

Thanks.

Addition:
Currently I log `self.losses `which seems to be wrong or?

`Losses: [array([-0.01116974,  0.01139735,  1.6872343 ], dtype=float32), array([-0.01316211,  0.00833996,  1.4614307 ], dtype=float32), array([-0.0151818 ,  0.00533725,  1.1907252 ], dtype=float32), array([-0.0078504 ,  0.00316278,  0.76215446], dtype=float32), array([nan, nan, nan], dtype=float32), array([nan, nan, nan], dtype=float32), array([nan, nan, nan], dtype=float`

**Comment by henrycharlesworth** [comment]
Hi!

I should have included a minimal example somewhere, but if you look at the "generateGUI.py" file it does load in network parameters (and gives you a GUI interface where you can play against the trained network). But a minimal example would be something like:

```from PPONetwork import PPONetwork
sess = tf.Session()
net = PPONetwork(sess, 412, 1695, "name") #412 is input size I think, 1695 is potential action space size
tf.global_variables_initializer().run(session=sess)
params = joblib.load("modelParameters...")
net.loadparams(params)
```

Then if you have a Big2Game created, something like:
```
_, state, avail_acs = game.getCurrentState()
action, value, neg_log_prob = net.step(state, availAcs)
reward, done, info = game.step(action[0])
```

However the fact you have NaNs in your loss would definitely suggest something is going wrong. Definitely try checking the parameter values (they may also be NaN - I did have some issues with this which I never fully resolved, but I thought I'd implemented a workaround). At the very least if you do this with the parameters included in the repo (modelParameters136500) the network should play a pretty solid game.

**Comment by CesMak** [comment]
Hey,
thanks for your fast response. I just saw the file `generateGUI.py`. In this everything is explained fine :)
Well your code works fine - but I adapted it for playing a much more easy game named [witches](https://github.com/CesMak/witches_ai). 

I just figured out, that in main2PPOSimulation.py (Line 188) loss is becoming NAN after 3 update.
```

loss_   = self.trainingModel.train(lrnow, cliprangenow, states[mb_inds], availAcs[mb_inds], returns[mb_inds], actions[mb_inds], values[mb_inds], neglogpacs[mb_inds])

loss_ is:
Update = 0
Loss: [2.0319765e-08, 0.020417742, 1.7447468]
Update >= 3
Loss: [nan, nan, nan]
```

For this reason the `needToReset = 1` is set and that is why my network does not learn anything it is reset always. 

I have two questions about the loss:
* What are the 3 Loss parameters?
* In my game a reward is given after each round is finished (4 steps with 4 players) and at the end the total result is the sum of all intermediate rewards. I think this is the main difference between my game and yours. Maybe you can give me a small hint where I should start debugging. I am not sure if I have to adjust the hyperparameters, the reward normalization factor or maybe the gamelogic is not correct.....

**Comment by henrycharlesworth** [comment]
OK cool, that sounds like an interesting project!

The three losses are the policy loss (if I remember correctly equation 7 in the PPO paper: https://arxiv.org/pdf/1707.06347.pdf), the value function loss (L^{VF} in equation 9 of that paper) and the average entropy over the available actions (also in the paper).

It's always hard to say how to go about debugging this kind of thing, especially with tensorflow (I actually never use tensorflow anymore, if only I had known about PyTorch when I started this project I think many hours of frustration would have been saved!). I would suspect that the hyperparameters are most likely OK - my understanding is that PPO is relatively robust to hyperparam choices (for a reinforcement learning algorithm, at least), and my choices more or less follow what's suggested in the original paper. For reward normalization I think a good rule of thumb would be to make each reward fall in the [-1,1] range, but I would also suspect this probably isn't the cause for the NaNs. One thing I would suggest checking in detail is how (or if) you're editing the run method of the big2PPOsimulation class, because my code here is really ugly and looking back now I can't 100% remember everything that's going on here in detail - but it is definitely big 2 specific. For example, one thing I do there which would be different for you is that I only assign rewards to the training minibatch when a game is finished (lines 108-114). I guess also just check the training batches that the run function is returning, and check that mb_rewards, mb_returns, mb_availAcs, mb_actions, mb_values, mb_neglogpacs etc all look like they're giving sensible values (and maybe compare them to a batch when running the big2 code, because these are the only things that the neural network training function ever sees). The one benefit you do have is that the error seems to be occurring right at the start of training, so hopefully it will be possible to figure out what's going wrong!

**Comment by CesMak** [comment]
Hey Henry,

wow your code is indeed hard to understand for me. So I have to assign the rewards every round (after 4 steps for 4 players). But it does actually not matter I also can assign them every step. However as far as I understood your code, you just asign the rewards to the active player (mb_pGos[-2][i]-1) 

I cannot figure out how to adapt these lines of the code (to give rewards every step) see mainbigPPOSimulation.py -> run:

```

  for _ in range(self.nSteps):
            currGos, currStates, currAvailAcs = self.vectorizedGame.getCurrStates()
            currStates = np.squeeze(currStates)
            currAvailAcs = np.squeeze(currAvailAcs)
            currGos = np.squeeze(currGos)
            actions, values, neglogpacs = self.trainingNetwork.step(currStates, currAvailAcs)
            rewards, dones, infos = self.vectorizedGame.step(actions)
            mb_obs.append(currStates.copy())
            mb_pGos.append(currGos)
            mb_availAcs.append(currAvailAcs.copy())
            mb_actions.append(actions)
            mb_values.append(values)
            mb_neglogpacs.append(neglogpacs)
            mb_dones.append(list(dones))
            #now back assign rewards if state is terminal
            toAppendRewards = np.zeros((self.nGames,))
            mb_rewards.append(toAppendRewards)
            for i in range(self.nGames):
                if dones[i] == True:
                    reward = rewards[i]
                    mb_rewards[-1][i] = reward[mb_pGos[-1][i]-1] / self.rewardNormalization
                    mb_rewards[-2][i] = reward[mb_pGos[-2][i]-1] / self.rewardNormalization
                    mb_rewards[-3][i] = reward[mb_pGos[-3][i]-1] / self.rewardNormalization
                    mb_rewards[-4][i] = reward[mb_pGos[-4][i]-1] / self.rewardNormalization
                    mb_dones[-2][i] = True
                    mb_dones[-3][i] = True
                    mb_dones[-4][i] = True
                    self.epInfos.append(infos[i])
                    self.gamesDone += 1
                    print("Game %d finished. Lasted %d turns" % (self.gamesDone, infos[i]['numTurns']))
        self.prevObs = mb_obs[endLength:]
        self.prevGos = mb_pGos[endLength:]
```

If you have some time you can also have a look at my [adopted version](https://github.com/CesMak/witches_ai/blob/master/PPO/mainBig2PPOSimulation.py) it still produces NAN simply run it with 

`witches_pytorch/PPO$ python mainBig2PPOSimulation.py `

Thanks for your help

**Comment by henrycharlesworth** [comment]
So I wrote this code over two years ago and I can't remember all of the details (for future projects I am going to actually write proper comments...). I think the way this works is as follows: mb_rewards (minibatch rewards) ends up being a timesteps x num_games array. The code takes advantage of the fact that in big two you always take turns sequentially, so p1 plays, then p2, then p3, then p4, then p1, etc etc until the game is over - so each players decisions are always 4 time steps apart. So mb_rewards[0][i], mb_rewards[4][i], mb_rewards[8][i], ... will all be rewards for a particular player in game i, mb_rewards[1][i], mb_rewards[5][i], ... will be for a different player etc. In big 2 the only rewards I apply are at the end of a game, when the last player plays their last turn - this means I have to retroactively assign rewards to the last decisions that each player made and tell them that it was the last decision they made (at the time they didn't know it was their last decision. So once we have made the final decision and find that dones[i] (for game i) is done, we take rewards[i], which is a vector of four reward values for p1, p2, p3, p4, and:
mb_rewards[-1][i] = reward[mb_pGos[-1][i]-1] ... says we assign the reward of the previous player in game i (mb_pGos[-1][i]), then
mb_rewards[-2][i] = reward[mb_pGos[-2][i]-1] ... says we assign the reward of the player before that in game i, etc - so we work backwards and retroactively apply the rewards. Having the mb_rewards set up in this way allows us to apply generalized advantage estimation to get the returns later on - in a way that we automatically are training on the experience of each of the four players separately.

Not sure if this is very clear, but hopefully it helps a bit!
